File: MapleStoryServer/ServerCore/RecvBuffer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ServerCore
{
    public class RecvBuffer
    {
        ArraySegment<byte> _buffer;

        int _readPos;
        int _writePos;

        public RecvBuffer(int bufferLength)
        {
            _buffer = new ArraySegment<byte>(new byte[bufferLength], 0, bufferLength);
        }

        public int ReadSize { get { return _writePos - _readPos; } }
        public int WriteSize { get { return _buffer.Count - _writePos; } }

        public ArraySegment<byte> ReadSegment { get { return new ArraySegment<byte>(_buffer.Array, _buffer.Offset + _readPos, ReadSize); } }
        public ArraySegment<byte> WriteSegment { get { return new ArraySegment<byte>(_buffer.Array, _buffer.Offset + _writePos, WriteSize); } }
        public void Clear()
        {
            int readSize = ReadSize;
            if (_readPos == _writePos)
            {
                _readPos = _writePos = 0;
            }
            else
            {
                Array.Copy(_buffer.Array, _buffer.Offset + _readPos, _buffer.Array, _buffer.Offset, readSize);
                _readPos = 0;
                _writePos = readSize;
            }
        }
        public bool OnRead(int numOfBytes)
        {
            if (ReadSize < numOfBytes)
                return false;
            _readPos += numOfBytes;
            return true;
        }

        public bool OnWrite(int numOfBytes)
        {
            if (WriteSize < numOfBytes)
                return false ;
            _writePos += numOfBytes;
            return true;
        }
    }
}

File: Server/ServerCore/RecvBuffer.cpp
#include "pch.h"
#include "RecvBuffer.h"

RecvBuffer::~RecvBuffer()
{
}

RecvBuffer::RecvBuffer(uint16 size)
{
	_buffer.assign(size, (byte)0);
}

uint16 RecvBuffer::GetReadSize()
{
	return _writePos - _readPos;
}

uint16 RecvBuffer::GetWriteSize()
{
	return _buffer.size() - _writePos;
}

byte* RecvBuffer::GetReadSegment()
{
	return &_buffer[_readPos];
}

byte* RecvBuffer::GetWriteSegment()
{
	return &_buffer[_writePos];
}

RecvBuffer::operator byte* ()
{
	return _buffer.data();
}

bool RecvBuffer::OnWrite(uint16 numOfBytes)
{
	if (numOfBytes > GetWriteSize())
		return false;
	_writePos += numOfBytes;
	return true;
}

bool RecvBuffer::OnRead(uint16 numOfBytes)
{
	if (numOfBytes > GetReadSize())
		return false;
	_readPos += numOfBytes;
	return true;
}

void RecvBuffer::Clear()
{
	uint16 readSize = GetReadSize();
	if (_readPos == _writePos)
	{
		_readPos = 0;
		_writePos = 0;
	}
	else
	{
		for (uint16 i = 0; i < readSize; i++)
			_buffer[i] = _buffer[i + _readPos];
		_readPos = 0;
		_writePos = readSize;
	}
}

void RecvBuffer::Init()
{
	_readPos = 0;
	_writePos = 0;
	uint32 size = _buffer.size();
	for (uint32 i = 0; i < size; i++)
		_buffer[i] = (byte)0;
}
